### PR TITLE
Debugging the supervised model.

### PR DESCRIPTION
**Pull request by oskopek**
Model runs now. I changed the NNs a bit, now they kind of perform a "pointer" across the board, if I made no mistake. Can't run on Leonhard, but runs locally for a few iterations. We need to add a way to predict and/or evaluate how well the model is performing.

### CODE REVIEW

File: codenames/supervised/network.py [hunk 1]
```diff
@@ -6,67 +6,93 @@
 import torch.nn as nn
 import torch.nn.functional as F
 
+from .data import BOARD_TENSOR_MAX_LENGTH
```
Comment (jendelel) [inline codenames/supervised/network.py]: What does .data mean?? Why not just data?

File: codenames/supervised/network.py [hunk 2]
```diff
@@ -6,67 +6,93 @@
...
+        self.attn_combine = nn.Linear(2 * enc_hidden_size, dec_hidden_size)
+        self.dropout = nn.Dropout(dropout_p)
+        self.gru = nn.GRU(dec_hidden_size, dec_hidden_size)
+        self.out = nn.Linear(dec_hidden_size, output_size)
+
+    def forward(self, embedded_inputs, dec_hidden, encoder_outputs):
+        """
+        Execute one time step of the decoder RNN.
+
+        :param embedded_inputs: Shape (1, batch, enc_hs).
```
Comment (jendelel) [inline codenames/supervised/network.py]: +1

File: codenames/supervised/network.py [hunk 3]
```diff
@@ -6,67 +6,93 @@
...
+
+        embedded_dropout = self.dropout(embedded_inputs)
+
+        attn_input = torch.cat((embedded_dropout, dec_hidden[0]), dim=-1)  # (batch, enc_hs + dec_hs)
+        attn_output = self.attn(attn_input)
+        attn_weights = F.softmax(attn_output, dim=-1).unsqueeze(0)  # (1, batch, max_len)
+
+        # batch-first:
+        enc_len = encoder_outputs.size(0)
+        attn_weights = attn_weights[:, :, :enc_len].transpose(0, 1)  # (batch, 1, enc_len)
```
Comment (jendelel) [inline codenames/supervised/network.py]: Do we need this? Can't we just swap the arguments in the bmm? I trust you that the attention is correct.

File: codenames/supervised/train.py [hunk 4]
```diff
@@ -156,22 +160,32 @@ def main():
...
+        'embedding_dim': hidden_size
+    })
+    clue_embed, _ = embeddings.setup_embeddings({
+        'embedding_path': args.embedding_path,
+        'num_embeddings': len(clue_vocab),
+        'embedding_dim': hidden_size
+    })
+
+    encoder = network.EncoderRNN(hidden_size).to(device)
+    clue_scaler = nn.Linear(len(clue_vocab), hidden_size).to(device)  # TODO(oskopek): Shouldn't this be an embedding?
```
Comment (jendelel) [inline codenames/supervised/train.py]: What do you mean with the TODO?

File: codenames/supervised/network.py (hunk 1 above)
Comment (oskopek) [reply codenames/supervised/network.py]: It's a relative import. The other variant would be `from codenames.supervised.data import BOARD_TENSOR_MAX_LENGTH`. I'll do whatever you prefer.

File: codenames/supervised/network.py (hunk 3 above)
Comment (oskopek) [reply codenames/supervised/network.py]: Well, BMM does batch matrix multiplication, and the first dimension is the batch (after the transpose). Not sure what you mean by swapping the arguments in BMM, that would work if it were transpose(1, 2), right?

File: codenames/supervised/train.py (hunk 4 above)
Comment (oskopek) [reply codenames/supervised/train.py]: As in, if we shouldn't use nn.Embedding for this somehow. I'm not actually sure here.

**Comment by oskopek** [comment]
Also, I just realized this doesn't work on CUDA for some reason (on Leonhard). I'll open a followup PR with the fixes for that, sometime in the future.

File: codenames/supervised/network.py (hunk 1 above)
Comment (jendelel) [reply codenames/supervised/network.py]: TIL. Thanks. Leave it :)

File: codenames/supervised/network.py (hunk 3 above)
Comment (jendelel) [reply codenames/supervised/network.py]: Sure. If it works...

File: codenames/supervised/train.py (hunk 4 above)
Comment (jendelel) [reply codenames/supervised/train.py]: ok

**Pull request merged**